doc/modules/neighbors.rst

```diff
@@ -415,14 +415,16 @@ depends on a number of factors:
   a significant fraction of the total cost.  If very few query points
   will be required, brute force is better than a tree-based method.
 
-Currently, ``algorithm = 'auto'`` selects ``'brute'`` if :math:`k >= N/2`,
-the input data is sparse, or ``effective_metric_`` isn't in
-the ``VALID_METRICS`` list for either ``'kd_tree'`` or ``'ball_tree'``.
-Otherwise, it selects the first out of ``'kd_tree'`` and ``'ball_tree'``
-that has ``effective_metric_`` in its ``VALID_METRICS`` list.
-This choice is based on the assumption that the number of query points is at
-least the same order as the number of training points, and that ``leaf_size``
-is close to its default value of ``30``.
+Currently, ``algorithm = 'auto'`` selects ``'brute'`` if
+``metric = 'precomputed'``, or :math:`D > 15`, or :math:`k >= N/2`, or the
+input data is sparse, or ``effective_metric_`` isn't in the ``VALID_METRICS``
+list for either ``'kd_tree'`` or ``'ball_tree'``. Otherwise, it selects the
+first out of ``'kd_tree'`` and ``'ball_tree'`` that has ``effective_metric_``
+in its ``VALID_METRICS`` list. This heuristic is based on the assumptions that
+the number of query points is at least the same order as the number of
+training points, that ``leaf_size`` is close to its default value of ``30``,
+and that when the number of features exceeds 15, the intrinsic dimensionality
+of the data is generally to high for tree-based methods.
```

Comment:
Maybe let's do bullet points for this sentence and the conditions above? As currently they are long sentences with lots of information packed that are hard to read.

"""
Currently, ``algorithm = 'auto'`` selects ``'brute'`` if any of the following conditions are verified,
- ``metric = 'precomputed'``
- or math:`D > 15`
...

Otherwise...

This heuristic is based on the following assumptions:
- the number of query points is at least
- ``leaf_size`` is close to its default value of ``30``
- when the number of features exceeds 15, ...
"""